File: pythonCode/colorTransferBetweenImage.py
import cv2
import numpy as np

#Read the images
source = cv2.imread("../assets/src.jpeg")
destination = cv2.imread("../assets/dest.jpeg")

#initialize result image
result = np.copy(destination)

#convert to LAB color space
sourceLab = cv2.cvtColor(source, cv2.COLOR_BGR2Lab)
destinationLab = cv2.cvtColor(destination, cv2.COLOR_BGR2LAB)
resultLab = cv2.cvtColor(result, cv2.COLOR_BGR2Lab)

#convert to float32
sourceLab = np.float32(sourceLab)
destinationLab = np.float32(destinationLab)
resultLab = np.float32(resultLab)

#split into channels
sourceL, sourceA, sourceB = cv2.split(sourceLab)
destinationL, destinationA, destinationB = cv2.split(destinationLab)
resultL, resultA, resultB = cv2.split(resultLab)

#subtract the mean of the destination. This in a way removes the base color of the destination image.
resultL = destinationL - destinationL.mean()
resultA = destinationA - destinationA.mean()
resultB = destinationB - destinationB.mean()

#scale the standard deviation of the destination 
resultL = resultL * (sourceL.std()/ destinationL.std())
resultA = resultA * (sourceA.std()/ destinationA.std())
resultB = resultB * (sourceB.std()/ destinationB.std())

#add mean of the source image
resultL = resultL + sourceL.mean()
resultA = resultA + sourceA.mean()
resultB = resultB + sourceB.mean()

#make sure all image values are in the range 0 and 255
resultL = np.clip(resultL, 0, 255)
resultA = np.clip(resultA, 0, 255)
resultB = np.clip(resultB, 0, 255)

#merge back the channels
resultLab = cv2.merge([resultL, resultA, resultB])

#convert back from float32
resultLab = np.uint8(resultLab)

#convert back to RGB color space
result =  cv2.cvtColor(resultLab, cv2.COLOR_LAB2BGR)

#create windows to display images
cv2.namedWindow("source", cv2.WINDOW_NORMAL)
cv2.namedWindow("destination", cv2.WINDOW_NORMAL)
cv2.namedWindow("result", cv2.WINDOW_NORMAL)

#display images
cv2.imshow("source", source)
cv2.imshow("destination", destination)
cv2.imshow("result", result)

#press esc to exit the program
cv2.waitKey(0)

#close all the opened windows
cv2.destroyAllWindows()

File: cPlusPlusCode/colorTransferBetweenImage.cpp
#include<opencv2/imgproc.hpp>
#include<opencv2/highgui.hpp>
#include<iostream>

using namespace std;
using namespace cv;

int main(){

    //Read the source and destination images
    Mat source = imread("../assets/src.jpeg");
    Mat destination = imread("../assets/dest.jpeg");

    //Result Mat instialized
    Mat result = destination.clone();

    //declare Mat variables to store images in LAB color space
    Mat sourceLab, destinationLab, resultLab;

    //Converting to LAB color space
    cvtColor(source, sourceLab, COLOR_BGR2Lab);
    cvtColor(destination, destinationLab, COLOR_BGR2Lab);
    cvtColor(result, resultLab, COLOR_BGR2Lab);

    //Convert to float32
    sourceLab.convertTo(sourceLab, CV_32F);
    destinationLab.convertTo(destinationLab, CV_32F);
    resultLab.convertTo(resultLab, CV_32F);

    //Declare vectors to store splitted channels
    vector<Mat> sourceChannels;
    vector<Mat> destinationChannels;
    vector<Mat> resultChannels;

    //spliting channels
    split(sourceLab, sourceChannels);
    split(destinationLab, destinationChannels);
    split(resultLab, resultChannels);

    //Declare variables to store means and standard deviations
    float mean1, mean2, mean3, mean4,mean5, mean6;
    float stdDev1, stdDev2, stdDev3, stdDev4, stdDev5, stdDev6;
    Scalar meanSource, meanDestination, stdDevSource, stdDevDestination;

    //Finding means and standard deviations for different channels for source image
    meanStdDev(sourceLab, meanSource, stdDevSource, Mat());
    mean1 = meanSource.val[0];
    mean2 = meanSource.val[1];
    mean3 = meanSource.val[2];
    stdDev1 = stdDevSource.val[0];
    stdDev2 = stdDevSource.val[1];
    stdDev3 = stdDevSource.val[2];

    //Finding means and standard deviations for different channels for destination image
    meanStdDev(destinationLab, meanDestination, stdDevDestination, Mat());
    mean4 = meanDestination.val[0];
    mean5 = meanDestination.val[1];
    mean6 = meanDestination.val[2];
    stdDev4 = stdDevDestination.val[0];
    stdDev5 = stdDevDestination.val[1];
    stdDev6 = stdDevDestination.val[2];

    //subtract the mean of destination image. This in a way removes the base color of the destination image.
    resultChannels[0] = destinationChannels[0] - mean4;
    resultChannels[1] = destinationChannels[1] - mean5;
    resultChannels[2] = destinationChannels[2] - mean6;

    //scale the standard deviation of the destination image
    resultChannels[0] = resultChannels[0] * (stdDev1/ stdDev4);
    resultChannels[1] = resultChannels[1] * (stdDev2/ stdDev5);
    resultChannels[2] = resultChannels[2] * (stdDev3/ stdDev6);

    //add the mean of source image to get the color
    resultChannels[0] = resultChannels[0] + mean1;
    resultChannels[1] = resultChannels[1] + mean2;
    resultChannels[2] = resultChannels[2] + mean3;

    //merge the channels
    merge(resultChannels, resultLab);

    //convert back from float32
    resultLab.convertTo(resultLab, CV_8UC3);

    //convert the image back BGR Color space
    cvtColor(resultLab, result, COLOR_Lab2BGR);

    //Create windows to display images
    namedWindow("source", WINDOW_NORMAL);
    namedWindow("destination", WINDOW_NORMAL);
    namedWindow("result", WINDOW_NORMAL);

    //display images
    imshow("source", source);
    imshow("destination", destination);
    imshow("result", result);

    //press esc to exit the program
    waitKey(0);

    //close all the opened windows
    destroyAllWindows();

    return 0;
}